### PR TITLE
move collections to channel listing

### PR DESCRIPTION
**Pull request by d-wysocki**
I want to merge this change because...Create collection channel listing

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

**Comment by patrys** [comment]
Task linked: [SALEOR-1229 Move collection availability to mulichannel solution](https://app.clickup.com/t/2549495/SALEOR-1229)

File: saleor/graphql/product/types/products.py [hunk 1]
```diff
@@ -797,17 +808,24 @@ class Collection(CountableDjangoObjectType):
     background_image = graphene.Field(
         Image, size=graphene.Int(description="Size of the image.")
     )
-    translation = TranslationField(CollectionTranslation, type_name="collection")
+    translation = TranslationField(
+        CollectionTranslation,
+        type_name="collection",
+        resolver=ChannelContextType.resolve_translation,
+    )
+    channel_listing = graphene.List(
```
Comment (maarcingebala) [inline saleor/graphql/product/types/products.py:816]: Not `channel_listings`? I see that we have the name `channel_listing`  which returns a list in more places on this branch, I guess we should change this name to plural everywhere @d-wysocki  @fowczarek? Also, the description here isn't very accurate. I would simply say "List of channels in which the collection is available".

File: saleor/product/error_codes.py [hunk 2]
```diff
@@ -18,3 +18,22 @@ class ProductErrorCode(Enum):
...
+    INVALID = "invalid"
+    PRODUCT_WITHOUT_CATEGORY = "product_without_category"
+    NOT_PRODUCTS_IMAGE = "not_products_image"
+    NOT_PRODUCTS_VARIANT = "not_products_variant"
+    NOT_FOUND = "not_found"
+    REQUIRED = "required"
+    UNIQUE = "unique"
+    VARIANT_NO_DIGITAL_CONTENT = "variant_no_digital_content"
+    CANNOT_MANAGE_PRODUCT_WITHOUT_VARIANT = "cannot_manage_product_without_variant"
+    PRODUCT_NOT_ASSIGNED_TO_CHANNEL = "product_not_assigned_to_channel"
```
Comment (maarcingebala) [inline saleor/product/error_codes.py:39]: Are these error codes relevant to collections? Not only these but also `ATRRIBUTE_ALREAD_ASSIGNED` and a few more. Can you re-check that?

File: saleor/graphql/product/types/products.py (hunk 1 above)
Comment (d-wysocki) [reply saleor/graphql/product/types/products.py:816]: Good point, I will fix this in other PR. For now I will leave a comment, sounds good?

Comment (maarcingebala) [reply saleor/graphql/product/types/products.py:816]: 👍

File: saleor/product/error_codes.py [hunk 3]
```diff
@@ -18,3 +18,13 @@ class ProductErrorCode(Enum):
...
+
+
+class CollectionErrorCode(Enum):
+    DUPLICATED_INPUT_ITEM = "duplicated_input_item"
+    GRAPHQL_ERROR = "graphql_error"
+    INVALID = "invalid"
+    NOT_FOUND = "not_found"
+    REQUIRED = "required"
+    UNIQUE = "unique"
+    CANNOT_MANAGE_PRODUCT_WITHOUT_VARIANT = "cannot_manage_product_without_variant"
```
Comment (IKarbowiak) [inline saleor/product/error_codes.py:30]: Do you use this error code somewhere? It doesn't sound like an error code for collection.

Comment (d-wysocki) [reply saleor/product/error_codes.py:30]: yes, its used in `CollectionAddProducts` which is related to collection errors

File: saleor/graphql/product/mutations/channels.py [hunk 4]
```diff
@@ -335,3 +346,76 @@ def perform_mutation(cls, _root, info, id, input):
...
+        error_type_class = CollectionChannelListingError
+        error_type_field = "collection_channel_listing_errors"
+
+    @classmethod
+    def add_channels(cls, collection: "CollectionModel", add_channels: List[Dict]):
+        for add_channel in add_channels:
+            defaults = {
+                "is_published": add_channel.get("is_published"),
+                "publication_date": add_channel.get("publication_date", None),
+            }
```
Comment (fowczarek) [inline saleor/graphql/product/mutations/channels.py:390]: If we implement it in this way we can accidentally remove some data. Look in `add_channels` on `ProductChannelListingUpdate`.

In your solution, if you update only `is_published` defaults will contain `"publication_date": None`, that means you want to clear this field.

File: saleor/graphql/product/mutations/products.py [hunk 5]
```diff
@@ -436,8 +456,8 @@ class Meta:
         description = "Update public metadata for collection."
         permissions = (ProductPermissions.MANAGE_PRODUCTS,)
         public = True
-        error_type_class = ProductError
-        error_type_field = "product_errors"
+        error_type_class = CollectionError
```
Comment (fowczarek) [inline saleor/graphql/product/mutations/products.py:459]: All mutations that return collection should wrap it in `ChannelContext`. 
You should add it to all deprecated collection meta mutations in this file.

File: saleor/product/models.py [hunk 6]
```diff
@@ -898,7 +895,36 @@ def get_ordering_queryset(self):
         return self.product.collectionproduct.all()
 
 
-class Collection(SeoModel, ModelWithMetadata, PublishableModel):
+class CollectionsQueryset(models.QuerySet):
```
Comment (fowczarek) [inline saleor/product/models.py:898]: Let's use the same method link in ProdcutsQueryset. If you want to implement a new convention for you should refactor all old approach.

File: saleor/graphql/product/resolvers.py [hunk 7]
```diff
@@ -25,16 +25,22 @@ def resolve_categories(info, level=None, **_kwargs):
     return qs.distinct()
 
 
-def resolve_collection_by_slug(info, slug):
+def resolve_collection_by_slug(info, slug, channel_slug):
     requestor = get_user_or_app_from_context(info.context)
     return (
-        models.Collection.objects.visible_to_user(requestor).filter(slug=slug).first()
+        models.Collection.objects.visible_to_user_with_channel_slug(
```
Comment (fowczarek) [inline saleor/graphql/product/resolvers.py:31]: In ProductsQueryset we use `visible_to_user` and it has a channel as param. 
We are not able to return `visible_to_user` for non-staff users without passing the channel as param.

File: saleor/graphql/product/tests/test_collection_channel_listing_update.py [hunk 8]
```diff
@@ -0,0 +1,466 @@
...
+    )
+    content = get_graphql_content(response)
+
+    # then
+    data = content["data"]["collectionChannelListingUpdate"]
+    collection_data = data["collection"]
+    assert not data["collectionChannelListingErrors"]
+    assert collection_data["slug"] == published_collection.slug
+    assert collection_data["channelListing"][0]["isPublished"] is False
+    assert collection_data["channelListing"][0]["publicationDate"] is None
```
Comment (fowczarek) [inline saleor/graphql/product/tests/test_collection_channel_listing_update.py:391]: Let's add a date to this fixture to find that we accidentally clear this value.

File: saleor/graphql/product/tests/test_collection_channel_listing_update.py [hunk 9]
```diff
@@ -0,0 +1,466 @@
...
+    publication_date = date.today()
+    collection_id = graphene.Node.to_global_id("Collection", collection.pk)
+    channel_id = graphene.Node.to_global_id("Channel", channel_USD.id)
+    variables = {
+        "id": collection_id,
+        "input": {
+            "addChannels": [
+                {
+                    "channelId": channel_id,
+                    "isPublished": True,
```
Comment (fowczarek) [inline saleor/graphql/product/tests/test_collection_channel_listing_update.py:408]: We should remove it. 
```suggestion
```

File: saleor/graphql/product/types/products.py [hunk 10]
```diff
@@ -638,8 +644,15 @@ def resolve_channel_listing(root: ChannelContext[models.Product], info, **_kwarg
...
-        return root.node.collections.all()
+    def resolve_collections(root: ChannelContext[models.Product], *_args, **_kwargs):
+        instances = root.node.collections.all()
+        channel_slug = root.channel_slug
+
+        collections = [
+            ChannelContext(node=collection, channel_slug=channel_slug)
+            for collection in instances
+        ]
+        return collections
```
Comment (fowczarek) [inline saleor/graphql/product/types/products.py:655]:
```suggestion
        return ChannelQsContext(qs=instances, channel_slug=channel_slug)
```

File: saleor/graphql/product/types/products.py [hunk 11]
```diff
@@ -842,14 +868,19 @@ def get_node(cls, info, id):
             return qs.filter(id=id).first()
         return None
 
+    @staticmethod
+    def resolve_channel_listing(root: ChannelContext[models.Collection], _info):
```
Comment (fowczarek) [inline saleor/graphql/product/types/products.py:872]: Lack of permissions.

File: saleor/product/error_codes.py (hunk 3 above)
Comment (fowczarek) [reply saleor/product/error_codes.py:30]: `collectionAddProducts` can return this error.

File: saleor/graphql/product/resolvers.py (hunk 7 above)
Comment (d-wysocki) [reply saleor/graphql/product/resolvers.py:31]: The changes are already implemented, I didn't commit changes because of extend this PR with removing homepage collection.

File: saleor/product/models.py (hunk 6 above)
Comment (d-wysocki) [reply saleor/product/models.py:898]: The changes are already implemented, I didn't commit changes because of extend this PR with removing homepage collection.

File: saleor/graphql/product/types/products.py (hunk 10 above)
Comment (d-wysocki) [reply saleor/graphql/product/types/products.py:655]: After your changes i got error `AssertionError: User Error: expected iterable, but did not find one for field Product.collections.`. I will stay with my solution.

File: saleor/graphql/meta/mutations.py [hunk 12]
```diff
@@ -118,7 +118,11 @@ def success_response(cls, instance):
         use_channel_context = any(
             [
                 isinstance(instance, Model)
-                for Model in [product_models.Product, product_models.ProductVariant]
+                for Model in [
+                    product_models.Product,
+                    product_models.ProductVariant,
+                    product_models.Collection,
```
[the remark this answers is not in the document]
Comment (d-wysocki) [reply saleor/graphql/meta/mutations.py:124]: `Sale`,`Voucher` and `ShippingMethod` are models without Meta, so we no need to add more models in Meta mutation.

File: saleor/graphql/product/filters.py [hunk 13]
```diff
@@ -391,17 +391,20 @@ class Meta:
 
 
 class CollectionFilter(django_filters.FilterSet):
-    published = EnumFilter(
-        input_class=CollectionPublished, method=filter_collection_publish
-    )
+    is_published = django_filters.BooleanFilter(method="filter_is_published")
```
[the remark this answers is not in the document]
Comment (d-wysocki) [reply saleor/graphql/product/filters.py:394]: I will stay with original solution, we can talk about this and change it in separate PR.

File: saleor/graphql/discount/tests/test_sale_sorting_with_channels.py [hunk 14]
```diff
@@ -133,7 +133,7 @@ def test_sales_with_sorting_and_channel_USD(
     sort_by["channel"] = channel_USD.slug
     variables = {"sortBy": sort_by}
 
-    # when
+    # whsaleor/graphql/shop/types.pyen
```
Comment (fowczarek) [inline saleor/graphql/discount/tests/test_sale_sorting_with_channels.py:136]: :police_officer:

File: saleor/graphql/core/tests/test_view.py [hunk 15]
```diff
@@ -271,32 +268,6 @@ def test_validation_errors_query_do_not_get_logged(
     assert graphql_log_handler.messages == []
 
 
-@mock.patch.object(Collection, "get_node")
```
Comment (fowczarek) [inline saleor/graphql/core/tests/test_view.py:274]: We shouldn't remove this test.

File: saleor/graphql/product/mutations/products.py [hunk 16]
```diff
@@ -471,8 +506,13 @@ class Meta:
...
+        error_type_field = "collection_errors"
+
+    @classmethod
+    def success_response(cls, instance):
+        instance = ChannelContext(node=instance, channel_slug=None)
+        return super().success_response(instance)
 
 
 class CategoryUpdateMeta(UpdateMetaBaseMutation):
```
Comment (fowczarek) [inline saleor/graphql/product/mutations/products.py:518]: Still unresolved:
"All mutations that return collection should wrap it in ChannelContext.
You should add it to all deprecated collection meta mutations in this file."

File: saleor/graphql/product/types/products.py (hunk 11 above)
Comment (fowczarek) [reply saleor/graphql/product/types/products.py:862]: Still unresolved. Non-staff users shouldn't be able to get any od channel_listing.

File: saleor/product/migrations/0135_collection_channellisting.py [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+        channel = channels_dict.get(currency)
+        if not channel:
+            channel, _ = Channel.objects.get_or_create(
+                currency_code=currency,
+                defaults={
+                    "name": f"Channel {currency}",
+                    "slug": f"channel-{currency.lower()}",
+                },
+            )
+            channels_dict[currency] = channel
```
Comment (fowczarek) [inline saleor/product/migrations/0135_collection_channellisting.py:26]: Maybe we should move to get the channel outside `for`.

File: saleor/static/populatedb_data.json [hunk 18]
```diff
@@ -10200,5 +10196,45 @@
...
+      "collection": 2,
+      "channel": 1
+    }
+  },
+  {
+    "model": "product.collectionchannellisting",
+    "pk": 3,
+    "fields": {
+      "publication_date": null,
+      "is_published": false,
```
Comment (fowczarek) [inline saleor/static/populatedb_data.json:10225]:
```suggestion
      "publication_date": "2020-01-01",
      "is_published": true,
```

**Review by fowczarek — APPROVED** [review]
:rocket:

**Review by maarcingebala — APPROVED** [review]
